File: lingting-spring-web/src/main/java/live/lingting/spring/web/wrapper/RepeatBodyRequestWrapper.java
package live.lingting.spring.web.wrapper;

import jakarta.servlet.ReadListener;
import jakarta.servlet.ServletInputStream;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletRequestWrapper;
import live.lingting.framework.util.FileUtils;
import live.lingting.framework.util.StreamUtils;
import lombok.Getter;

import java.io.BufferedReader;
import java.io.Closeable;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * @author lingting 2024-03-20 15:08
 */
public class RepeatBodyRequestWrapper extends HttpServletRequestWrapper implements Closeable {

	public static final File TMP_DIR = FileUtils.createTempDir("request");

	@Getter
	private final File bodyFile;

	private final Map<String, String[]> paramsMap;

	@Getter
	private final List<Closeable> closeableList;

	public RepeatBodyRequestWrapper(HttpServletRequest request) throws IOException {
		super(request);
		paramsMap = request.getParameterMap();
		bodyFile = FileUtils.createTemp(".repeat", TMP_DIR);
		try (FileOutputStream outputStream = new FileOutputStream(bodyFile)) {
			StreamUtils.write(request.getInputStream(), outputStream);
		}
		closeableList = new ArrayList<>();
	}

	public static RepeatBodyRequestWrapper of(HttpServletRequest request) throws IOException {
		if (!(request instanceof RepeatBodyRequestWrapper)) {
			return new RepeatBodyRequestWrapper(request);
		}
		return (RepeatBodyRequestWrapper) request;
	}

	@Override
	public BufferedReader getReader() throws IOException {
		BufferedReader bufferedReader = new BufferedReader(new FileReader(bodyFile));
		closeableList.add(bufferedReader);
		return bufferedReader;
	}

	@Override
	@SuppressWarnings("java:S2095")
	public ServletInputStream getInputStream() throws IOException {
		FileInputStream stream = new FileInputStream(bodyFile);
		ServletInputStream servletInputStream = new ServletInputStream() {
			@Override
			public boolean isFinished() {
				return false;
			}

			@Override
			public boolean isReady() {
				return false;
			}

			@Override
			public void setReadListener(ReadListener readListener) {
				//
			}

			@Override
			public int read() throws IOException {
				return stream.read();
			}

			@Override
			public void close() throws IOException {
				stream.close();
			}
		};
		closeableList.add(servletInputStream);
		return servletInputStream;
	}

	@Override
	public Map<String, String[]> getParameterMap() {
		return paramsMap;
	}

	@Override
	public void close() throws IOException {
		for (Closeable closeable : getCloseableList()) {
			StreamUtils.close(closeable);
		}
		FileUtils.delete(getBodyFile());
	}

}

File: lingting-spring-web/src/main/kotlin/live/lingting/spring/web/wrapper/RepeatBodyRequestWrapper.kt
package live.lingting.spring.web.wrapper

import jakarta.servlet.ReadListener
import jakarta.servlet.ServletInputStream
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletRequestWrapper
import live.lingting.framework.util.FileUtils
import live.lingting.framework.util.StreamUtils
import java.io.BufferedReader
import java.io.Closeable
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.FileReader

/**
 * @author lingting 2024-03-20 15:08
 */
class RepeatBodyRequestWrapper(request: HttpServletRequest) : HttpServletRequestWrapper(request), Closeable {

    companion object {

        @JvmField
        val TMP_DIR = FileUtils.createTempDir("request")

        @JvmStatic
        fun of(request: HttpServletRequest): RepeatBodyRequestWrapper {

            if (request !is RepeatBodyRequestWrapper) {
                return RepeatBodyRequestWrapper(request)
            }
            return request
        }

    }

    val bodyFile: File = FileUtils.createTemp(".repeat", TMP_DIR)

    private val paramsMap: MutableMap<String, Array<String>> = request.parameterMap

    val closeableList: MutableList<Closeable>

    init {
        FileOutputStream(bodyFile).use { out ->
            val input = request.inputStream
            StreamUtils.write(input, out)
        }
        closeableList = ArrayList<Closeable>()
    }

    override fun getReader(): BufferedReader {
        val bufferedReader = BufferedReader(FileReader(bodyFile))
        closeableList.add(bufferedReader)
        return bufferedReader
    }

    override fun getInputStream(): ServletInputStream {
        val stream = FileInputStream(bodyFile)
        val servletInputStream: ServletInputStream = object : ServletInputStream() {
            override fun isFinished(): Boolean {
                return false
            }

            override fun isReady(): Boolean {
                return false
            }

            override fun setReadListener(readListener: ReadListener) {
                //
            }

            override fun read(): Int {
                return stream.read()
            }

            override fun close() {
                stream.close()
            }
        }
        closeableList.add(servletInputStream)
        return servletInputStream
    }

    override fun getParameterMap(): MutableMap<String, Array<String>> {
        return paramsMap
    }

    override fun close() {
        for (closeable in this.closeableList) {
            StreamUtils.close(closeable)
        }
        FileUtils.delete(this.bodyFile)
    }

}
